omnichain/staking/contracts/Staking.sol

```diff
@@ -0,0 +1,102 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.7;
+
+import "@zetachain/protocol-contracts/contracts/zevm/SystemContract.sol";
+import "@zetachain/protocol-contracts/contracts/zevm/interfaces/zContract.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@zetachain/toolkit/contracts/BytesHelperLib.sol";
+
+contract Staking is ERC20, zContract {
+    error SenderNotSystemContract();
+    error WrongChain();
+    error NotAuthorizedToClaim();
+
+    SystemContract public immutable systemContract;
+    uint256 public immutable chain;
+
+    mapping(address => uint256) public stakes;
+    mapping(address => address) public beneficiaries;
+    mapping(address => uint256) public lastStakeTime;
+    uint256 public rewardRate = 1;
+
+    constructor(
+        string memory name,
+        string memory symbol,
+        uint256 chainID,
+        address systemContractAddress
+    ) ERC20(name, symbol) {
+        systemContract = SystemContract(systemContractAddress);
+        chain = chainID;
+    }
+
+    function onCrossChainCall(
+        zContext calldata context,
+        address zrc20,
+        uint256 amount,
+        bytes calldata message
+    ) external override {
+        if (msg.sender != address(systemContract)) {
```

Comment:
is more elegant a modifier instead of the direct validation

Reply:
This is a part of the template. I've created an issue for that and will update in the template first, then this will propagate to the examples: https://github.com/zeta-chain/template/issues/26

Reply:
I propose we mark this as resolved and I will update the examples once this is fixed in the template.